Issue templates

jgonggrijp opened the pull request:
I decided to try my hand at GitHub's issue template functionality. The changes can be previewed [at my fork](https://github.com/jgonggrijp/backbone/issues/new/choose), where I temporarily enabled issues. This is probably a better way to review this PR than by reading the YAML files. The top three categories have forms that you can view by clicking "get started".

I had two main goals:

1. Steer people away from reporting security vulnerabilities through public issue tickets. Those should be communicated privately to me and Jeremy instead.
2. Guide inexperienced users to document their issues as clearly as possible.

For expert users, there is still the option to skip the forms and create a blank issue.

I may have gone a bit overboard with the forms. I'm open to all suggestions.

jgonggrijp commented:
Bump: I would prefer to see opinions from at least two different people (excluding myself), before taking a decision. To anyone considering to do a review: you can skip the code, the preview link in the opening post will show you what it will actually look like in the GitHub web interface.

CC @paulfalgout @RayRaz @GammaGames

GammaGames commented:
I figured out how to preview them, they look good! The bug one might be a little long though, entering all that information would take a while. I left a few comments where I thought they could be combined without missing anything

paulfalgout reviewed (approved):
Fancy

Rayraz commented:
The bug form is quite long. I suppose things like logs, research, prior discussions, related tickets, etc. could all be wrapped into simply a `remarks` field?

Steps to reproduce and example code could possibly be merged into one field as well.

jgonggrijp commented:
> I figured out how to preview them, they look good! The bug one might be a little long though, entering all that information would take a while. I left a few comments where I thought they could be combined without missing anything

@GammaGames where can I see your comments?

jgonggrijp commented:
@GammaGames Ah, it says "pending" in that screenshot, which means they are probably only visible to you. GitHub queues your comments and then publishes them all at once when you submit your "overall" review (i.e., the "verdict"). You can make that final step by going to the "Files changed" tab and pressing the green button in the top right.

GammaGames commented on .github/ISSUE_TEMPLATE/Bugs.yml:
```diff
@@ -0,0 +1,231 @@
+  - type: markdown
+    attributes:
+      value: "
+        Thank you for taking the effort to report a bug.\n\n
```
Would literal style work here? For consistency, I see it's used later

jgonggrijp replied:
No, the heading hashes `##` would not work in that case.

GammaGames commented:
@jgonggrijp Thank you, posted! Guess I've accidentally done it correctly the few times I've posted a review

jgonggrijp commented on .github/ISSUE_TEMPLATE/Bugs.yml:
```diff
@@ -0,0 +1,231 @@
+    attributes:
+      label: Affected engine(s)
+      description: >
+        In which browser(s) or other JavaScript environment(s) did you encounter
```
I don't think we do, but that is an excellent idea. I'll see whether it is easy to add, and otherwise create a ticket for it.

GammaGames replied:
I think putting it all in one commit makes sense and at a glance it looked good! I'll try to take a closer look and respond to the rest of your comments sometime this weekend

jgonggrijp commented on .github/ISSUE_TEMPLATE/Bugs.yml:
No, please do get into that now. Why do you think expected and actual should be the same field?

jgonggrijp replied:
As it currently stands, about half of the questions in the bug issue template can be skipped. Would that reduce the prep sufficiently, you think? I think the required part of the form may even be a bit shorter than the fomantic template (although that template has no way to enforce any part in particular).

GammaGames replied:
I think the current required questions are a good minimum! You can mark this comment as resolved

jgonggrijp commented on .github/ISSUE_TEMPLATE/Features.yml:
```diff
@@ -0,0 +1,82 @@
+    id: goal
+    attributes:
+      label: Ultimate goal
+      description: >-
```
It is the same as `>` except that it removes trailing linebreaks. My editor inserts the `-` automatically when I type `>`, and at some point I ceased my resistance. I guess it is a bit tidier.

jgonggrijp commented on .github/ISSUE_TEMPLATE/Bugs.yml:
Just to make sure we are talking about the same thing: did you interpret this question as asking for a permalink to Backbone code, or for a permalink to someone's project code in which the bug manifested?

GammaGames replied:
I interpreted it as a link to the documentation, I may have misread it. Most repos have a jsfiddle template set up and just ask for a reproducable example

jgonggrijp commented on .github/ISSUE_TEMPLATE/Documentation.yml:
```diff
@@ -0,0 +1,79 @@
+    id: problem
+    attributes:
+      label: Problem statement
+      description: >-
```
I somewhat disagree here, it's asking for the effects rather than the cause. I know people generally find it more intuitive to discuss the cause before the effect, but it is the effect that actually motivates changes to the documentation. I find it important to start with the "why".

That said, your comment made me realize the naming of the field was a bit vague, so I renamed it to "effect".

GammaGames replied:
I agree with the name change

jgonggrijp commented on .github/ISSUE_TEMPLATE/Bugs.yml:
🤷

jgonggrijp commented:
I can only choose between single-line `<input>` and multi-line `<textarea>`. In the latter case, it is the same size as all other "large" boxes. Which of these options would you prefer?

- [ ] Shorten the placeholder so it fits.
- [ ] Turn the behavior fields into `<textarea>`s.

By they way, do you think that the [code size of the `debugInfo` function](https://github.com/jashkenas/backbone/pull/4278/commits/8fe1bbbb60ae7f66d7c3bed02f0b90d052fcb9ab) is proportional to its usefulness?

paulfalgout commented:
@jgonggrijp honestly I'd move that `debugInfo` into a separate directory/module such that people can
```js
import Backbone from 'backbone';
import { debugInfo } from 'backbone/debug-info';
```

GammaGames commented:
I like that they're `<input>`s, I think a shorter example would fit the space and help guide the user. And yes, the added code for `debugInfo` has a lot of value! Gathering all the versions as well as environment would be tedous without it

jgonggrijp commented:
Alright, I shortened the placeholder for that `<input>`, moved `debugInfo` to a separate module and added documentation for it.

The updated bug report template can be previewed [here](https://github.com/jgonggrijp/backbone/issues/new?template=Bugs.yml) and the documentation for `debugInfo` can be previewed [here](https://cdn.statically.io/gh/jgonggrijp/backbone/issue-templates/index.html#Utility-Backbone-debugInfo). I would appreciate a final thumbs-up before merging this!

(The linter currently chokes on the ESM syntax in `modules/debug-info.js`, I'll adress that before merging.)

GammaGames commented:
👍️
